clients/src/main/java/org/apache/kafka/common/feature/SupportedVersionRange.java

```diff
@@ -17,9 +17,16 @@
 package org.apache.kafka.common.feature;
 
 import java.util.Map;
+import java.util.Objects;
+import org.apache.kafka.common.utils.Utils;
 
 /**
- * An extended {@link BaseVersionRange} representing the min/max versions for supported features.
+ * An extended {@link BaseVersionRange} representing the min, max and first active versions for a
+ * supported feature:
+ *  - minVersion: This is the minimum supported version for the feature.
+ *  - maxVersion: This the maximum supported version for the feature.
+ *  - firstActiveVersion: This is the first active version for the feature. Versions in the range
```

Comment:
Thinking about this a bit more. It seems that the intention of firstActiveVersion is to avoid deploying a wrong version of the broker that causes the deprecation of a finalized feature version unexpectedly. However, the same mistake can happen with firstActiveVersion since the deprecation of a finalized feature version is based on firstActiveVersion. So, I am not sure if firstActiveVersion addresses a real problem.

In general, we tend to deprecate a version very slowly in AK. So, if the mistake is to deploy a new release that actually deprecates a supported version. Old clients are likely all gone. So, moving finalized min version to supported min version may not cause a big problem. We can just document that people should make sure old versions are no longer used before deploying new releases.

If the mistake is to deploy an old version of the broker whose maxSupportedVersion is < maxFinalizedVersion, we will fail the broker. So, this mistake can be prevented.

Reply:
@junrao :

I'd like to discuss an example that cites a problem I'm concerned about.

> In general, we tend to deprecate a version very slowly in AK. So, if the mistake is to deploy a new release that actually deprecates a supported version. Old clients are likely all gone. So, moving finalized min version to supported min version may not cause a big problem. We can just document that people should make sure old versions are no longer used before deploying new releases.

Let's say we have some feature `F` whose:
 * Supported version range is: `[minVersion=1, maxVersion=6]`
 * Existing finalized version range in the cluster is: `[minVersionLevel=1, maxVersionLevel=6]`

Now, let us say a point in time arrives when we need to deprecate the feature version `1`.
Let us say we bump up supported `minVersion` to `2` in a subsequent major Kafka release.
Before this new release is deployed, let us assume the cluster operator knows 100% that old clients that were using the feature at version `1` are gone, so this is not a problem.

**PROBLEM:** Still, if we deploy this new release, the broker will consider the following as a feature version incompatibility.
 * Supported version range is: `[minVersion=2, maxVersion=6]`
 * Existing finalized version range in the cluster is: `[minVersionLevel=1, maxVersionLevel=6]`

Upon startup of a broker thats using the new release binary, the above combination will crash the broker since supported `minVersion=2` is greater than `minVersionLevel=1`. Basically the versioning system thinks that there is now a broker that does not support `minVersionLevel=1`, which does not adhere to the rules of the system.

Here is my thought: This is where `firstActiveVersion` becomes useful. By bumping it up during a release (instead of the supported feature's `minVersion`), we are able to get past this situation. When `firstActiveVersion`is advanced in the code, and the cluster is deployed, the controller (and all brokers) know that the advancement acts a request to the controller to act upon the feature deprecation (by writing the advanced value to the `FeatureZNode`). So, in this case we would release the broker with the supported feature version range: `[minVersion=1, firstActiveVersion=2, maxVersion=6]`, and the broker release wouldn't fail (because the intent is clearly expressed to the versioning system).

What are your thoughts on the above?

Reply:
@kowshik : I was thinking what if we relax the current check by just making sure that maxVersion of finalized is within the supported range. Basically in your example, if supported minVersion goes to 2, it's still allowed since it's less than maxVersion of finalized. However, if supported minVersion goes to 7, this fails the broker since it's more than maxVersion of finalized.

Your concern for the relaxed check seems to be around deploying a wrong version of the broker by mistake. I am not sure if that's a big concern. If the wrong broker affects maxVersion of finalized, the broker won't start. If the wrong broker affects minVersion of finalized, if we deprecated slowly, it won't impact the existing clients.

Reply:
@junrao My response below.

> I was thinking what if we relax the current check by just making sure that maxVersion of finalized is within the supported range. Basically in your example, if supported minVersion goes to 2, it's still allowed since it's less than maxVersion of finalized. However, if supported minVersion goes to 7, this fails the broker since it's more than maxVersion of finalized.

If we relaxed the current check like you suggested, then, as of today the following will happen:
The controller can not effectively finalize `minVersionLevel` for the feature, because, with a relaxed check we do not know whether all brokers in the cluster support a particular `minVersion` when the controller finalizes it at a particular value.

Supposing we want to keep the concept of `minVersionLevel` like the way it is now (i.e. it tells you the lowest feature version level guaranteed to be supported by any broker in the cluster), then, `firstActiveVersion` provides the door to mutate it safely whenever the need arises.

So when thinking about your suggestion, it is not clear to me if you are intending to alter the meaning of `minVersionLevel`? If so, please could you tell me what is the new meaning of `minVersionLevel` that you are proposing?

> Your concern for the relaxed check seems to be around deploying a wrong version of the broker by mistake.

Not exactly, my concern is a bit different. It is to be noted that `minVersionLevel` can not be changed through the `UPDATE_FEATURES` api. Assuming we want to maintain the meaning of `minVersionLevel` the way it is today, my concern is about how would one alter `minVersionLevel` for a feature when the need arises? This needs to be done without letting the versioning system misconstrue the intent with a feature version incompatibility. This is where usage of `firstActiveVersion` acts as a solution.

Reply:
"we do not know whether all brokers in the cluster support a particular minVersion when the controller finalizes the minVersionLevel at a particular value." The controller knows the minSupportedVersion for all brokers, right? What if we do the following? When finalizing a feature, the controllers uses the highest minSupportedVersion across all brokers as finalizedMinVersion, as long as it's <= finalizedMaxVersion. On broker restart, we also advance finalizedMinVersion if the new broker's minSupportedVersion has advanced (assuming still <= finalizedMaxVersion).

Reply:
@junrao Awesome. This is a very good point. The approach you proposed is very elegant, and we should shoot for it, when we’re giving the benefit of the doubt on deprecation to the broker binary version. I’ll update the KIP with details and share with community for feedback. As soon as that is done, I'll follow up in separate PR implementing this logic.